bmSettings.lua

```diff
@@ -34,6 +38,369 @@ BMSettings.Constants.UpdateRates = {
+                                db:exec("COMMIT")
+                                db:exec("BEGIN TRANSACTION")
+                            end
+                        end
+                    end
+                end
+            end
+            db:exec("COMMIT")
+        end
+        db:close()
+        print("Conversion complete!")
+    end
+end
+
+-- Retrieve and Deserialize Data
+function BMSettings:retrieveDataFromDB()
+    local db = self:InitializeDB()
+    self.settings = {
+        Global = {},
+        Version = 8,
+        LastBackup = os.time(),
+        Sets = {},
+        Buttons = {},
+        Characters = {},
+    }
+
+    local globalSettingsData = self:loadFromDB(db, "SELECT settings_version, settings_last_backup FROM settings WHERE server='global' AND character='global'")
+    if globalSettingsData[1] then
+        -- self.settings.Global = {
+        --     ButtonSize = globalSettingsData[1].settings_button_size,
+        -- }
+        self.settings.Version = globalSettingsData[1].settings_version
+        self.settings.LastBackup = globalSettingsData[1].settings_last_backup
+    end
+
+    local setsData = self:loadFromDB(db, "SELECT set_name, button_number, button_id FROM sets")
+    for _, set in ipairs(setsData) do
+        self.settings.Sets[set.set_name] = self.settings.Sets[set.set_name] or {}
+        self.settings.Sets[set.set_name][set.button_number] = set.button_id
+    end
+
+    local buttonsData = self:loadFromDB(db, "SELECT * FROM buttons")
+    for _, button in ipairs(buttonsData) do
+        self.settings.Buttons[button.button_number] = {
+            Label = button.button_label,
+            highestRenderTime = button.button_render,
+            TextColorRGB = button.button_text_color ~= "" and button.button_text_color or nil,
+            ButtonColorRGB = button.button_button_color ~= "" and button.button_button_color or nil,
+            CachedCountDown = button.button_cached_countdown,
+            CachedCoolDownTimer = button.button_cached_cooldown,
+            CachedToggleLocked = button.button_cached_toggle_locked,
+            CachedLastRan = button.button_cached_last_run,
+            labelMidX = button.button_label_mid_x,
+            labelMidY = button.button_label_mid_y,
+            CachedLabel = button.button_cached_label,
+            Cmd = button.button_cmd,
+            EvaluateLabel = button.button_evaluate_label == 1,
+            ShowLabel = button.button_show_label == 1,
+            Icon = tonumber(button.button_icon) > 0 and tonumber(button.button_icon) or nil,
+            IconType = button.button_icon_type ~= "" and button.button_icon_type or "Item",
+            IconLua = button.button_icon_lua,
+            TimerType = button.button_timer_type,
+            Cooldown = button.button_cooldown,
+        }
+    end
+
+    local charactersData = self:loadFromDB(db, "SELECT character, character_locked, character_hide_title FROM characters")

```

Comment:
Can we avoid pre-loading every characters data now that we have moved to a db?

Reply:
yeah that we can do. i would still pull the char names for local copy functionality but yeah no reason to have all char data loaded on one character